### PR TITLE
Add initial support for constants in python API

### PR DESCRIPTION
**Pull request by klecki**
#### Why we need this PR?
It allows to use constants with arithmetic operators.

#### What happened in this PR?
 - Explain solution of the problem, new feature added.
int is treated as int32
float is treated as float32

To indicate type of constant operand nvidia.dali.types.Constant
was introduced
 - What was changed, added, removed?
 - What is most important part that reviewers should focus on?
 - Was this PR tested? How?
 - Were docs and examples updated, if necessary?

**JIRA TASK**: [DALI-1107]

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/ops.py [hunk 1]
```diff
@@ -601,29 +601,103 @@ def _choose_device(inputs):
...
+            DALIDataType.UINT16: "uint16",
+            DALIDataType.UINT32: "uint32",
+            DALIDataType.UINT64: "uint64",
+            DALIDataType.FLOAT16: "float16",
+            DALIDataType.FLOAT: "float32",
+            DALIDataType.FLOAT64: "float64",
+        }
+        return dtype_to_desc[input.dtype]
+    raise TypeError(("Argument to arithmetic operation not supported. Got {}, expected "
+                "a return value from other DALI Operator or a constant value of type 'int', "
```
Comment (JanuszL) [inline dali/python/nvidia/dali/ops.py]: `a return value from other DALI Operator` - but I don't see any check against it here?

File: dali/python/nvidia/dali/ops.py [hunk 2]
```diff
@@ -601,29 +601,103 @@ def _choose_device(inputs):
...
+            raise TypeError(("Argument to arithmetic operation not supported. Got {}, expected "
+                    "a return value from other DALI Operator  or a constant value of type 'int', "
+                    "'float' or 'nvidia.dali.Constant'.").format(str(type(input))))
+    if len(integers) == 0:
+        integers = None
+    if len(reals) == 0:
+        reals = None
+    return (categories_idxs, edges, integers, reals)
+
+def _generate_input_desc(categories_idx, integers, reals):
```
Comment (JanuszL) [inline dali/python/nvidia/dali/ops.py]: I would put some comment what is happening here - like building an expression with some syntax. There is one description in the native code, but probably the python side is the first that user will check looking for details.

File: dali/python/nvidia/dali/ops.py [hunk 3]
```diff
@@ -601,29 +601,103 @@ def _choose_device(inputs):
...
+                DALIDataType.INT64, DALIDataType.UINT8, DALIDataType.UINT16, DALIDataType.UINT32,
+                DALIDataType.UINT64]:
+            return True
+    return False
+
+def _is_real_like(input):
+    if type(input) == float:
+        return True
+    if isinstance(input, Constant):
+        if input.dtype in [DALIDataType.FLOAT16, DALIDataType.FLOAT, DALIDataType.FLOAT64]:
```
Comment (JanuszL) [inline dali/python/nvidia/dali/ops.py]: Can't you use just `_float_types` from types.py?

File: dali/python/nvidia/dali/ops.py [hunk 4]
```diff
@@ -601,29 +601,103 @@ def _choose_device(inputs):
...
-        # Create "instance" of operator
-        op = ArithmeticGenericOp(device = _choose_device(inputs), expression_desc = expression_desc)
-        # If we are on gpu, we must mark all inputs as gpu
-        if dev == "gpu":
-            dev_inputs = list(input.gpu() for input in inputs)
+def _is_integer_like(input):
+    if type(input) == int:
+        return True
+    if isinstance(input, Constant):
+        if input.dtype in [DALIDataType.INT8, DALIDataType.INT16, DALIDataType.INT32,
```
Comment (JanuszL) [inline dali/python/nvidia/dali/ops.py]: Can't you use just `_int_types` from types.py?

File: dali/python/nvidia/dali/types.py [hunk 5]
```diff
@@ -101,3 +101,89 @@ def __init__(self, ptr=0):
...
+        return self.value == other.value and self.dtype == other.dtype
+
+    def __int__(self):
+        if self.dtype in _int_types:
+            return self.value
+        raise TypeError("DALI Constant must be converted to one of int types explicitly before "
+                "casting to builtin 'int'.")
+
+    def __float__(self):
+        if self.dtype in [DALIDataType.FLOAT16, DALIDataType.FLOAT, DALIDataType.FLOAT64]:
```
Comment (JanuszL) [inline dali/python/nvidia/dali/types.py]:
```suggestion
        if self.dtype in _float_types:
``
?

File: dali/test/python/test_operator_arithmetic_ops.py [hunk 6]
```diff
@@ -28,12 +29,30 @@
...
 shape = (8, 8)
 
-devices = ["cpu_cpu", "cpu_gpu", "gpu_gpu", "gpu_gpu"]
+# A number used to test constant inputs
+magic_number = 42
 
-# float16 is marked as TODO in backend for gpu
-types = [np.int8, np.int16, np.int32, np.int64, np.uint8, np.uint16, np.uint32, np.uint64,
-         np.float32, np.float64]
+# We cannot have constant x constant operations with DALI op.
```
Comment (JanuszL) [inline dali/test/python/test_operator_arithmetic_ops.py]:
```suggestion
# We cannot have constant * constant operations with DALI op.
```

File: dali/test/python/test_ops_internals.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+# Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (JanuszL) [inline dali/test/python/test_ops_internals.py]: You need to rename it to `test_operator_*` to be picked by the wildcard in test script, or add it manually to test.sh in the appropriate TLX_ folder.

File: dali/python/nvidia/dali/ops.py (hunk 4 above)
Comment (klecki) [reply dali/python/nvidia/dali/ops.py]: Done.

File: dali/python/nvidia/dali/ops.py (hunk 3 above)
Comment (klecki) [reply dali/python/nvidia/dali/ops.py]: Done

File: dali/python/nvidia/dali/ops.py (hunk 1 above)
Comment (klecki) [reply dali/python/nvidia/dali/ops.py]: Copy-paste. Done.

File: dali/python/nvidia/dali/ops.py (hunk 2 above)
Comment (klecki) [reply dali/python/nvidia/dali/ops.py]: The user should not be looking for it here, as this is an internal function and he is not to create such description by hand.
I added a comment pointing to ArithmeticGenericOp.

File: dali/python/nvidia/dali/types.py (hunk 5 above)
Comment (klecki) [reply dali/python/nvidia/dali/types.py]: Done, missed it.

File: dali/test/python/test_operator_arithmetic_ops.py (hunk 6 above)
Comment (klecki) [reply dali/test/python/test_operator_arithmetic_ops.py]: `*` suggest multiplication.

File: dali/test/python/test_ops_internals.py (hunk 7 above)
Comment (klecki) [reply dali/test/python/test_ops_internals.py]: Done

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/ops.py [hunk 8]
```diff
@@ -601,29 +601,108 @@ def _choose_device(inputs):
...
+            categories_idxs.append(("real", len(reals)))
+            reals.append(input)
         else:
-            dev_inputs = inputs
-        # Call it imediatelly
-        return op(*dev_inputs)
-
-
+            raise TypeError(("Argument to arithmetic operation not supported. Got {}, expected "
+                    "a return value from other DALI Operator  or a constant value of type 'int', "
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/ops.py]:
```suggestion
                    "a return value from other DALI Operator or a constant value of type 'int', "
```

File: dali/python/nvidia/dali/ops.py [hunk 9]
```diff
@@ -601,29 +601,108 @@ def _choose_device(inputs):
...
+    if len(integers) == 0:
+        integers = None
+    if len(reals) == 0:
+        reals = None
+    return (categories_idxs, edges, integers, reals)
+
+# Generate the list of <input> subexpression as specified
+# by grammar for ArithmeticGenericOp
+def _generate_input_desc(categories_idx, integers, reals):
+    input_desc = ""
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/ops.py]: nitpick idea: You could do `input_desc = []` and append strings to it. At the end you do `" ".join(input_desc)`

Comment (klecki) [reply dali/python/nvidia/dali/ops.py]: Cool idea, will do it in a follow-up.

File: dali/python/nvidia/dali/ops.py (hunk 2 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/ops.py]: I know, but still someone may wan to check internals what is not that uncommon for python code.

**Pull request merged**